### PR TITLE
Cusparse assert

### PR DESCRIPTION
**Pull request by Rombur**
Add `AssertCusparse` to check that the call  to a cuSPARSE function was successful. Print the error message instead of the error code.

### CODE REVIEW

File: include/deal.II/base/exceptions.h [hunk 1]
```diff
@@ -314,6 +319,74 @@ namespace deal_II_exceptions
                               const char       *cond,
                               const char       *exc_name,
                               ExceptionBase     e) noexcept;
+#ifdef DEAL_II_WITH_CUDA
+    /**
+     * Return a string given an error code. This is similar to the cudaGetErrorString
+     * function but there is no equivalent function for cuSPARSE.
+     */
+    inline
+    std::string get_cusparse_error_string(cusparseStatus_t error_code)
```
Comment (bangerth) [inline include/deal.II/base/exceptions.h]: can you declare the function here and move its implementation into the .cc file?

Comment (Rombur) [reply include/deal.II/base/exceptions.h]: Sure, I don't know why I thought that there was no .cc file.

**Comment by Rombur** [comment]
Done

File: include/deal.II/base/exceptions.h [hunk 2]
```diff
@@ -314,6 +319,13 @@ namespace deal_II_exceptions
                               const char       *cond,
                               const char       *exc_name,
                               ExceptionBase     e) noexcept;
+#ifdef DEAL_II_WITH_CUDA
+    /**
+     * Return a string given an error code. This is similar to the cudaGetErrorString
+     * function but there is no equivalent function for cuSPARSE.
+     */
+    std::string get_cusparse_error_string(cusparseStatus_t error_code);
```
Comment (bangerth) [inline include/deal.II/base/exceptions.h]: make the argument `const` as per our usual convention

File: source/base/exceptions.cc [hunk 3]
```diff
@@ -484,6 +484,72 @@ namespace deal_II_exceptions
         }
     }
 
+
+
+#ifdef DEAL_II_WITH_CUDA
+    std::string get_cusparse_error_string(cusparseStatus_t error_code)
```
Comment (bangerth) [inline source/base/exceptions.cc]: same here

File: source/base/exceptions.cc [hunk 4]
```diff
@@ -484,6 +484,72 @@ namespace deal_II_exceptions
...
+    {
+      std::string message;
+      switch (error_code)
+        {
+        case CUSPARSE_STATUS_NOT_INITIALIZED:
+        {
+          message = "The cuSPARSE library was not initialized";
+
+          break;
```
Comment (bangerth) [inline source/base/exceptions.cc]: you could just do `return "...";` here and in all of the other places. saves you two lines each :-)

**Pull request merged**